db/db_impl/compacted_db_impl.cc

```diff
@@ -40,16 +40,35 @@ size_t CompactedDBImpl::FindFile(const Slice& key) {
 
 Status CompactedDBImpl::Get(const ReadOptions& options, ColumnFamilyHandle*,
                             const Slice& key, PinnableSlice* value) {
+  return Get(options, nullptr, key, value, /*timestamp*/ nullptr);
+}
+
+Status CompactedDBImpl::Get(const ReadOptions& options, ColumnFamilyHandle*,
+                            const Slice& key, PinnableSlice* value,
+                            std::string* timestamp) {
   assert(user_comparator_);
-  if (options.timestamp || user_comparator_->timestamp_size()) {
-    // TODO: support timestamp
-    return Status::NotSupported();
+  if (options.timestamp) {
+    const Status s =
+        FailIfTsSizesMismatch(DefaultColumnFamily(), *(options.timestamp));
+    if (!s.ok()) {
+      return s;
+    }
+  } else {
+    const Status s = FailIfCfHasTs(DefaultColumnFamily());
+    if (!s.ok()) {
+      return s;
+    }
   }
+  GetWithTimestampReadCallback read_cb(kMaxSequenceNumber);
+  std::string* ts =
+      user_comparator_->timestamp_size() > 0 ? timestamp : nullptr;
+  LookupKey lkey(key, kMaxSequenceNumber, options.timestamp);
   GetContext get_context(user_comparator_, nullptr, nullptr, nullptr,
-                         GetContext::kNotFound, key, value, nullptr, nullptr,
-                         nullptr, true, nullptr, nullptr);
-  LookupKey lkey(key, kMaxSequenceNumber);
-  Status s = files_.files[FindFile(key)].fd.table_reader->Get(
+                         GetContext::kNotFound, lkey.user_key(), value, ts,
+                         nullptr, nullptr, true, nullptr, nullptr, nullptr,
+                         nullptr, &read_cb);
+
+  Status s = files_.files[FindFile(lkey.user_key())].fd.table_reader->Get(
```

Comment:
Nit: a similar optimization can be done here
```
if (user_comparator_->CompareWithoutTimestamp(key, /*a_has_ts=*/false, ExtractUserKeyAndStripTimestamp(f.smallest_key), /*b_has_ts=*/false))) {
  // no need to search
}
```
Example as below. Suppose you search for 'b'
```
[('a', 100)    ('a', 90)]  [('c', 80), ('c', 70)]
```

Also please add a test to cover this case.

Reply:
Thank you for the suggestion, added the optimization.
I added a test for failing the sanity check when write with timestamps and read without timestamps, did the same for the MultiGet tests.

Reply:
Added a test case for this with `CheckDBOpenedAsCompactedDBWithOnlyLmaxFiles` check.